### DISCUSSION TITLE
Enabling HMR for additional pages ?

### DISCUSSION THREAD

**Discussion by stevebeauge** [aklinker1/vite-plugin-web-extension#116] [Q&A]
I'm writing an extension which inject buttons on top of some inputs, in order to inject business data.

The extension is using a content script to extend the page's DOM and opens an overlay whenever the custom button is clicked.

The overlay consists in an IFrame that targets an html file from the extension.

To sum up, I have in my extension:
* popup.html : (basic GUI for the extension button where the user can configure the extension)
* content.ts : content script that seek for <input> and add the button
* modal.html : content of the Iframe

The button in the content script will create on the fly the iframe element and set its src to the modal.html like this:

````
        const iframe = document.createElement("iframe");
        iframe.width = "800px";
        iframe.height = "600px";
        iframe.style.margin = "auto";
        const popup = browser.runtime.getURL('src/content/modal.html');
        iframe.setAttribute("src", popup);
        overlay.appendChild(iframe);
````

In order to have the html page bundled, I updated the vite config like this:

````
export default defineConfig({
  plugins: [
    react(),
    webExtension({
      manifest: generateManifest,
      additionalInputs: ['src/content/modal.html'],
    }),
  ],
  resolve: {
    alias: {
      // In dev mode, make sure fast refresh works
      "/@react-refresh": path.resolve(
        "node_modules/@vitejs/plugin-react-swc/refresh-runtime.js"
      ),
    },
  },
});
````

This is working as expected, my modal is properly handled by vite.

However, there's not hot reloading applied within the modal. Everytime I update something, I can see vite's web server recompiling the page, but nothing is updated within the browser (no HMR). I have to manually refresh the extension to see changes.

As far as I understand the source code of this plugin, the observed behavior is normal since this code:

https://github.com/aklinker1/vite-plugin-web-extension/blob/883460dade7892a6d1919edda1956b2efaddc9df/packages/vite-plugin-web-extension/src/build/getViteConfigsForInputs.ts#L101C1-L111

only take the entry points in the HMR process, not the additional files.

Is there anything that can be done ?

**Comment by stevebeauge** [comment]
I created a repro : https://github.com/stevebeauge/repro-vite-browser-ext-no-hmr-in-modal

**Comment by aklinker1** [comment]
> As far as I understand the source code of this plugin, the observed behavior is normal since this code:
>
> https://github.com/aklinker1/vite-plugin-web-extension/blob/883460dade7892a6d1919edda1956b2efaddc9df/packages/vite-plugin-web-extension/src/build/getViteConfigsForInputs.ts#L101C1-L111

HTML files listed under `additionalInputs` are passed into the multipage build you're referencing here. At least they should. Checking your repro now.

https://github.com/aklinker1/vite-plugin-web-extension/blob/883460dade7892a6d1919edda1956b2efaddc9df/packages/vite-plugin-web-extension/src/build/getViteConfigsForInputs.ts#L203-L217

**Comment by aklinker1** [comment]
So a couple of things I've noticed about the `@vitejs/plugin-react-swc` plugin. It is very touchy about how components should be written. If your components don't follow these rules, it doesn't work as of now.

1. Components exported by default
2. The component needs to be an anonymous function, not a named function or an arrow function
3. One component per file

So you need to change your `pages/Modal.tsx` component like so:

```diff
-const Modal: React.FC = () => {
+export default function() {
    return (
        <div>
            <h1>
                Hello from modal!
            </h1>
        </div>
    );
};
-
-export { Modal };
```

And it works. But that's not a great solution, also not the standard way of writing react, though I am out of date in my react knowledge.

I've verified that the regular vite + react + swc template (from `pnpm create vite`) doesn't have these requirements. I'll look into it, though it may be related to #114. HMR also seems touchy with the old `@vitejs/plugin-react`... so switching to that won't fix the problem.
